subsys/bluetooth/mesh/transport.c

```diff
@@ -910,36 +920,37 @@ static int trans_ack(struct bt_mesh_net_rx *rx, uint8_t hdr,
 	}
 
 	if (new_seg_ack) {
-		tx->attempts_left_without_progress =
-			BT_MESH_SAR_TX_RETRANS_NO_PROGRESS;
+		tx->attempts_left_without_progress = BT_MESH_SAR_TX_RETRANS_NO_PROGRESS;
+		tx->ack_received = 1U;
 	}
 
 	if (tx->nack_count) {
-		/* If transmission is not in progress it means
-		 * that Retransmission Timer is running
+
+		/* If there is an ongoing transmission that has not finished,
+		 * there are no more retransmit attempts left, or the retransmit
+		 * timer is not running, then we return early.
 		 */
-		if (new_seg_ack) {
-			if (tx->seg_o == 0) {
-				uint32_t delta_ms = (uint32_t)(k_uptime_get() -
-							       tx->adv_start_timestamp);
-				k_timeout_t timeout = K_NO_WAIT;
-
-				/* According to the Bluetooth Mesh Profile specification,
-				 * section 3.5.3.3, we should reset the retransmit timer and
-				 * retransmit immediately when receiving a valid ack message
-				 * while Retransmisison timer is running. However, transport should
-				 * still keep segment transmission interval time between
-				 * transmission of each segment.
-				 */
-				if (delta_ms < BT_MESH_SAR_TX_SEG_INT_MS) {
-					timeout = K_MSEC(BT_MESH_SAR_TX_SEG_INT_MS - delta_ms);
-				}
+		if (tx->seg_o != 0 || !seg_retrans_attempts_check(tx) ||
+		    !k_work_delayable_is_pending(&tx->retransmit)) {
```

Comment:
I'm a bit uncertain about `!seg_retrans_attempts_check(tx) || !k_work_delayable_is_pending(&tx->retransmit)` part.

The `tx->retransmit` timer starts in 3 places: here, in `schedule_transmit_continue` and in `seg_tx_send_unacked`.

While `seg_o != 0`, the algorithm is the following. When `seg_tx_send_unacked` is called for the first time and until all segments are sent out to the air, the `tx->retransmit` timer is used for sending the segments and it is rescheduled by response from the advertiser using `schedule_transmit_continue` function (see `seg_sent_cb` callbacks).

When the last segment was sent, `seg_tx_send_unacked` gets called again (`seg_sent_cb.end` -> `schedule_transmit_continue` -> `seg_tx_send_unacked`). Then `seg_o` is set to 0 and this time `tx.retransmit` works as the SAR Unicast Retransmission Timer.

If at the last step (when `seg_o` is set to 0), the `ack_received` is set to 1, then `seg_tx_send_unacked` will reschedule `tx->retransmit` timer immediately and will start resending the unack-ed segments. If `ack_received` was 0, then retransmission timeout will be used.

So, IOW, while `seg_o` is NOT 0, no need to restart the timer (so the first check in this if-statement is needed). But, if `seg_o` is 0, then the SAR Unicast Retransmission Timer is running and we need to reschedule it from here. `ack_received` can be reset to 0 in this case, as new segments will anyway trigger segacks from the receiver.

So I think what needs to be changed is that the if-statement on the line 921 in the old implementation with `if (new_seg_ack) {` should be removed (since the trigger for SAR Unicast Retransmission Timer restart is now any segack, not only new). Then, if `seg_o` is 0, we need to start retransmitting the segments, otherwise, just set `ack_received` flag and let `seg_tx_send_unacked` do the job.

Reply:
Ok, so you think that `tx->seg_o != 0` is sufficient assurance that the retransmission timer is pending, and `k_work_delayable_is_pending(&tx->retransmit)` can be skipped. What about `seg_retrans_attempts_check(tx)` then? I guess that it could be removed and it work fine. However, it bugs me a bit to let it stay as it was since as far as I can see we might end up waiting for as long as BT_MESH_SAR_TX_SEG_INT_MS before ending the transfer even if there are no more retransmit attempts. In reality I do not think this makes much of a difference, but when trying to read the code against spec it is hard to verify that the implementation is following spec since `seg_tx_send_unacked` is used as a Swiss army knife.

Reply:
I will try to revise it 👍

Reply:
> so you think that `tx->seg_o != 0` is sufficient assurance that the retransmission timer is pending, and `k_work_delayable_is_pending(&tx->retransmit)` can be skipped

Yes.

> we might end up waiting for as long as BT_MESH_SAR_TX_SEG_INT_MS before ending the transfer even if there are no more retransmit attempts

I think you are right here.